src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -89,39 +138,44 @@ private class CloseRevokedPartitions { }
         public IActorRef ConsumerActor { get; private set; } = null!;
 
         public PromiseControl<(TopicPartition, Source<TMessage, NotUsed>)> Control { get; }
-
-        /// <summary>
-        /// SubSourceLogic
-        /// </summary>
-        public SubSourceLogic(SourceShape<(TopicPartition, Source<TMessage, NotUsed>)> shape, ConsumerSettings<K, V> settings,
-                              IAutoSubscription subscription, Func<SubSourceLogic<K, V, TMessage>, IMessageBuilder<K, V, TMessage>> messageBuilderFactory,
-                              Option<Func<IImmutableSet<TopicPartition>, Task<IImmutableSet<TopicPartitionOffset>>>> getOffsetsOnAssign,
-                              Action<IImmutableSet<TopicPartition>> onRevoke, Attributes attributes)
+        
+        public SubSourceLogic(SourceShape<(TopicPartition, Source<TMessage, NotUsed>)> shape,
+            ConsumerSettings<K, V> settings,
+            IAutoSubscription subscription,
+            Option<Func<IImmutableSet<TopicPartition>, Task<IImmutableSet<TopicPartitionOffset>>>> getOffsetsOnAssign,
+            Action<IImmutableSet<TopicPartition>> onRevoke,
+            ISubSourceStageLogicFactory<K, V, TMessage> subSourceStageLogicFactory,
+            Attributes attributes)
             : base(shape)
         {
             _shape = shape;
             _settings = settings;
             _subscription = subscription;
-            _messageBuilder = messageBuilderFactory(this);
```

Comment:
the `IMessageBuilder` is now and internal implementation detail of the sub-stages, not the parent